linux/driver/osp-iio.c

```diff
@@ -0,0 +1,1203 @@
+				break;
+			default:
+				break;
+			}
+			ret = IIO_VAL_INT;
+			break;
+		
+		default:
+			break;
+		}
+		break;
+	default:
+		break;
+	}
+	mutex_unlock(&st->lock);
+	return ret;
+}
+
+/**
+ * osp_sensor_write_raw() - data write function.
+ * @indio_dev:	the struct iio_dev associated with this device instance
+ * @chan:	the channel whose data is to be read
+ * @val:	first element of returned value (typically INT)
+ * @val2:	second element of returned value (typically MICRO)
+ * @mask:	what we actually want to read. 0 is the channel, everything else
+ *		is as per the info_mask in iio_chan_spec.
+ *
+ * Note that all raw writes are assumed IIO_VAL_INT and info mask elements
+ * are assumed to be IIO_INT_PLUS_MICRO unless the callback write_raw_get_fmt
+ * in struct iio_info is provided by the driver.
+ */
+static int osp_sensor_write_raw(struct iio_dev *indio_dev,
+			       struct iio_chan_spec const *chan,
+			       int val,
+			       int val2,
+			       long mask)
+{
+#if 0
+	int i;
+	int ret = 0;
+	struct iio_dummy_state *st = iio_priv(indio_dev);
+
+	switch (mask) {
+	case 0:
+		if (chan->output == 0)
+			return -EINVAL;
+
+		/* Locking not required as writing single value */
+		mutex_lock(&st->lock);
+		st->dac_val = val;
+		mutex_unlock(&st->lock);
+		return 0;
+	case IIO_CHAN_INFO_CALIBBIAS:
+		mutex_lock(&st->lock);
+		/* Compare against table - hard matching here */
+		for (i = 0; i < ARRAY_SIZE(dummy_scales); i++)
+			if (val == dummy_scales[i].val &&
+			    val2 == dummy_scales[i].val2)
+				break;
+		if (i == ARRAY_SIZE(dummy_scales))
+			ret = -EINVAL;
+		else
+			st->accel_calibscale = &dummy_scales[i];
+		mutex_unlock(&st->lock);
+		return ret;
+	default:
+		return -EINVAL;
+	}
+#else
+	return 0;
+#endif
+}
+
+/**
+ * osp_iio_sensor_init() - Sets up sensor specific details.
+ * @indio_dev: the iio device structure
+ *
+ * Most drivers have one of these to set up default values,
+ * reset the device to known state etc.
+ *
+ * Init data to known values for debugging.
+ */
+static int osp_iio_sensor_init(struct iio_dev *indio_dev, int sensor, int type)
+{
+	struct osp_iio_sensor *osp_sensor = iio_priv(indio_dev);
+
+	osp_sensor->indio_dev = indio_dev;
+	memset(&osp_sensor->data, 0, sizeof(osp_sensor->data));
+	osp_sensor->sensor = sensor;
+
+	if (type == 0) {
+		switch (sensor) {
+		case SENSOR_ACCELEROMETER:
+			osp_sensor->data.xyz.x = 0;
+			osp_sensor->data.xyz.y = 1;
+			osp_sensor->data.xyz.z = 2;
+			break;
+		case SENSOR_GYROSCOPE:
+			osp_sensor->data.xyz.x = 2002;
+			osp_sensor->data.xyz.y = 2003;
+			osp_sensor->data.xyz.z = 2004;
+			break;
+		case SENSOR_MAGNETIC_FIELD:
+			osp_sensor->data.xyz.x = 4002;
+			osp_sensor->data.xyz.y = 4003;
+			osp_sensor->data.xyz.z = 4004;
+			break;
+		case SENSOR_ROTATION_VECTOR:
+			osp_sensor->data.quat.x = 6001;
+			osp_sensor->data.quat.y = 6002;
+			osp_sensor->data.quat.z = 6003;
+			osp_sensor->data.quat.r = 1004;
+			break;
+		}
+	}
+
```

Comment:
Total No of android sensors created : 16 ..
While creating IIO device, we are initializing data for android sensors  : SENSOR_ACCELEROMETER, SENSOR_GYROSCOPE, SENSOR_MAGNETIC_FIELD &
SENSOR_ROTATION_VECTOR. do we need to set data for other sensors as well?